### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [2]:
# Dependencies and Setup
import pandas as pd
# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"
# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)
# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [4]:
# calculate total number of schools
total_schools = len(school_data_complete["school_name"].unique())
# calculate total number of students
total_students = len(school_data_complete["student_name"])
# calculate total budget
total_budget = sum(school_data_complete["budget"].unique())
# calculate average math scores
avg_math_score = school_data_complete["math_score"].mean()
# calculate average reading scores
avg_reading_score = school_data_complete["reading_score"].mean()
# calculate percentage of students who received score of 70 or higher in math
passing_math_score = (school_data_complete["math_score"] >= 70).mean()
# calculate percentage of students who received score of 70 or higher in reading
passing_reading_score = (school_data_complete["reading_score"] >= 70).mean()
# calculate percentage of students who received score of 70 or higher in both math and reading
overall_passing = ((school_data_complete["math_score"] >= 70) & \
                   (school_data_complete["reading_score"] >= 70)).mean()
# creat new dataframe
district_summary = pd.DataFrame({
    "Total Schools": [total_schools],
    "Total Students": total_students,
    "Total Budget": total_budget,
    "Avg Math Score": avg_math_score,
    "Avg Reading Score": avg_reading_score,
    "% Passing Math": passing_math_score,
    "% Passing Reading": passing_reading_score,
    "% Overall Passing": overall_passing})
# format data frame for cleaner look
district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,}".format)
district_summary["Avg Math Score"] = district_summary["Avg Math Score"].map("{:.2f}".format)
district_summary["Avg Reading Score"] = district_summary["Avg Reading Score"].map("{:.2f}".format)
district_summary["% Passing Math"] = district_summary["% Passing Math"].map("{:.2%}".format)
district_summary["% Passing Reading"] = district_summary["% Passing Reading"].map("{:.2%}".format)
district_summary["% Overall Passing"] = district_summary["% Overall Passing"].map("{:.2%}".format)

district_summary

Total Schools Total Students Total Budget Avg Math Score Avg Reading Score  \
0             15         39,170  $24,649,428          78.99             81.88   

  % Passing Math % Passing Reading % Overall Passing  
0         74.98%            85.81%            65.17%

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

In [5]:
# find different schools
school_list = school_data_complete.groupby(["school_name"])
# list school types
type_of_school = school_data.set_index("school_name")["type"]
# calculate total number of students for each school
kids_per_school = school_list["Student ID"].count()
# calculate total school budget
school_budget = school_data.set_index("school_name")["budget"]
# calculate school budget per student
budget_per_student = school_budget / kids_per_school
# calculate average math scores
avg_math_score_sch = school_list["math_score"].mean()
# calculate average reading scores
avg_read_score_sch = school_list["reading_score"].mean()
# calculate percentage of students passing math
passing_math_sch = school_data_complete.loc[school_data_complete["math_score"] \
                >= 70].groupby("school_name")["Student ID"].count() / kids_per_school
# calculate percentage of students passing reading
passing_read_sch = school_data_complete.loc[school_data_complete["reading_score"] \
                >= 70].groupby("school_name")["Student ID"].count() / kids_per_school
# calculate percentage of students who passed math AND reading
overall_pass_both = school_data_complete.loc[(school_data_complete["math_score"] >= 70) & \
            (school_data_complete["reading_score"] >= 70)].groupby("school_name")\
            ["Student ID"].count() / kids_per_school
# create dataframe
school_summary = pd.DataFrame({
    "School Type": type_of_school,
    "Total Students": kids_per_school,
    "Total School Budget": school_budget,
    "Per Student Budget": budget_per_student,
    "Avg Math Score": avg_math_score_sch,
    "Avg Reading Score": avg_read_score_sch,
    "% Passing Math": passing_math_sch,
    "% Passing Reading": passing_read_sch,
    "% Overall Passing": overall_pass_both})
# copying the dataframe to preserve school_summary dataframe for later use
school_summary_df = school_summary.copy()
# format dataframe for cleaner look
school_summary_df["Total Students"] = school_summary_df["Total Students"].map("{:,}".format)
school_summary_df["Total School Budget"] = school_summary_df["Total School Budget"].map("${:,}".format)
school_summary_df["Per Student Budget"] = school_summary_df["Per Student Budget"].map("${:,.0f}".format)
school_summary_df["Avg Math Score"] = school_summary_df["Avg Math Score"].map("{:.2f}".format)
school_summary_df["Avg Reading Score"] = school_summary_df["Avg Reading Score"].map("{:.2f}".format)
school_summary_df["% Passing Math"] = school_summary_df["% Passing Math"].map("{:.2%}".format)
school_summary_df["% Passing Reading"] = school_summary_df["% Passing Reading"].map("{:.2%}".format)
school_summary_df["% Overall Passing"] = school_summary_df["% Overall Passing"].map("{:.2%}".format)

school_summary_df

School Type Total Students Total School Budget  \
Bailey High School       District          4,976          $3,124,928   
Cabrera High School       Charter          1,858          $1,081,356   
Figueroa High School     District          2,949          $1,884,411   
Ford High School         District          2,739          $1,763,916   
Griffin High School       Charter          1,468            $917,500   
Hernandez High School    District          4,635          $3,022,020   
Holden High School        Charter            427            $248,087   
Huang High School        District          2,917          $1,910,635   
Johnson High School      District          4,761          $3,094,650   
Pena High School          Charter            962            $585,858   
Rodriguez High School    District          3,999          $2,547,363   
Shelton High School       Charter          1,761          $1,056,600   
Thomas High School        Charter          1,635          $1,043,130   
Wilson High School        Charter          2,283          $1,319,574   
Wright High School        Charter          1,800          $1,049,400   

                      Per Student Budget Avg Math Score Avg Reading Score  \
Bailey High School                  $628          77.05             81.03   
Cabrera High School                 $582          83.06             83.98   
Figueroa High School                $639          76.71             81.16   
Ford High School                    $644          77.10             80.75   
Griffin High School                 $625          83.35             83.82   
Hernandez High School               $652          77.29             80.93   
Holden High School                  $581          83.80             83.81   
Huang High School                   $655          76.63             81.18   
Johnson High School                 $650          77.07             80.97   
Pena High School                    $609          83.84             84.04   
Rodriguez High School               $637          76.84             80.74   
Shelton High School                 $600          83.36             83.73   
Thomas High School                  $638          83.42             83.85   
Wilson High School                  $578          83.27             83.99   
Wright High School                  $583          83.68             83.95   

                      % Passing Math % Passing Reading % Overall Passing  
Bailey High School            66.68%            81.93%            54.64%  
Cabrera High School           94.13%            97.04%            91.33%  
Figueroa High School          65.99%            80.74%            53.20%  
Ford High School              68.31%            79.30%            54.29%  
Griffin High School           93.39%            97.14%            90.60%  
Hernandez High School         66.75%            80.86%            53.53%  
Holden High School            92.51%            96.25%            89.23%  
Huang High School             65.68%            81.32%            53.51%  
Johnson High School           66.06%            81.22%            53.54%  
Pena High School              94.59%            95.95%            90.54%  
Rodriguez High School         66.37%            80.22%            52.99%  
Shelton High School           93.87%            95.85%            89.89%  
Thomas High School            93.27%            97.31%            90.95%  
Wilson High School            93.87%            96.54%            90.58%  
Wright High School            93.33%            96.61%            90.33%

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [6]:
# sort dataframe by '% Overall Passing' to find top 5 performers
top_schools = school_summary_df.sort_values(["% Overall Passing"], ascending=False)
# print the top 5 schools
top_schools = top_schools.head(5)
top_schools

School Type Total Students Total School Budget  \
Cabrera High School     Charter          1,858          $1,081,356   
Thomas High School      Charter          1,635          $1,043,130   
Griffin High School     Charter          1,468            $917,500   
Wilson High School      Charter          2,283          $1,319,574   
Pena High School        Charter            962            $585,858   

                    Per Student Budget Avg Math Score Avg Reading Score  \
Cabrera High School               $582          83.06             83.98   
Thomas High School                $638          83.42             83.85   
Griffin High School               $625          83.35             83.82   
Wilson High School                $578          83.27             83.99   
Pena High School                  $609          83.84             84.04   

                    % Passing Math % Passing Reading % Overall Passing  
Cabrera High School         94.13%            97.04%            91.33%  
Thomas High School          93.27%            97.31%            90.95%  
Griffin High School         93.39%            97.14%            90.60%  
Wilson High School          93.87%            96.54%            90.58%  
Pena High School            94.59%            95.95%            90.54%

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [7]:
# sort dataframe by '% Overall Passing' to find bottom 5 performers
bottom_schools = school_summary_df.sort_values(["% Overall Passing"], ascending=False)
# print the bottom 5 schools
bottom_schools = bottom_schools.tail(5)
bottom_schools

School Type Total Students Total School Budget  \
Johnson High School      District          4,761          $3,094,650   
Hernandez High School    District          4,635          $3,022,020   
Huang High School        District          2,917          $1,910,635   
Figueroa High School     District          2,949          $1,884,411   
Rodriguez High School    District          3,999          $2,547,363   

                      Per Student Budget Avg Math Score Avg Reading Score  \
Johnson High School                 $650          77.07             80.97   
Hernandez High School               $652          77.29             80.93   
Huang High School                   $655          76.63             81.18   
Figueroa High School                $639          76.71             81.16   
Rodriguez High School               $637          76.84             80.74   

                      % Passing Math % Passing Reading % Overall Passing  
Johnson High School           66.06%            81.22%            53.54%  
Hernandez High School         66.75%            80.86%            53.53%  
Huang High School             65.68%            81.32%            53.51%  
Figueroa High School          65.99%            80.74%            53.20%  
Rodriguez High School         66.37%            80.22%            52.99%

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [8]:
# find average math score per school for 9th grade
grade9_math = student_data.loc[student_data["grade"] == "9th"].groupby("school_name")\
                ["math_score"].mean()
# find average math score per school for 10th grade
grade10_math = student_data.loc[student_data["grade"] == "10th"].groupby("school_name")\
                ["math_score"].mean()
# find average math score per school for 11th grade
grade11_math = student_data.loc[student_data["grade"] == "11th"].groupby("school_name")\
                ["math_score"].mean()
# find average math score per school for 12th grade
grade12_math = student_data.loc[student_data["grade"] == "12th"].groupby("school_name")\
                ["math_score"].mean()

# create dataframe
math_scores_grade = pd.DataFrame({
    "9th": grade9_math,
    "10th": grade10_math,
    "11th": grade11_math,
    "12th": grade12_math})

# format dataframe for cleaner look
math_scores_grade["9th"] = math_scores_grade["9th"].map("{:.2f}".format)
math_scores_grade["10th"] = math_scores_grade["10th"].map("{:.2f}".format)
math_scores_grade["11th"] = math_scores_grade["11th"].map("{:.2f}".format)
math_scores_grade["12th"] = math_scores_grade["12th"].map("{:.2f}".format)
math_scores_grade.index.name = None

math_scores_grade

9th   10th   11th   12th
Bailey High School     77.08  77.00  77.52  76.49
Cabrera High School    83.09  83.15  82.77  83.28
Figueroa High School   76.40  76.54  76.88  77.15
Ford High School       77.36  77.67  76.92  76.18
Griffin High School    82.04  84.23  83.84  83.36
Hernandez High School  77.44  77.34  77.14  77.19
Holden High School     83.79  83.43  85.00  82.86
Huang High School      77.03  75.91  76.45  77.23
Johnson High School    77.19  76.69  77.49  76.86
Pena High School       83.63  83.37  84.33  84.12
Rodriguez High School  76.86  76.61  76.40  77.69
Shelton High School    83.42  82.92  83.38  83.78
Thomas High School     83.59  83.09  83.50  83.50
Wilson High School     83.09  83.72  83.20  83.04
Wright High School     83.26  84.01  83.84  83.64

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [9]:
# find average reading score per school for 9th grade
grade9_read = student_data.loc[student_data["grade"] == "9th"].groupby("school_name")\
                ["reading_score"].mean()
# find average reading score per school for 10th grade
grade10_read = student_data.loc[student_data["grade"] == "10th"].groupby("school_name")\
                ["reading_score"].mean()
# find average reading score per school for 11th grade
grade11_read = student_data.loc[student_data["grade"] == "11th"].groupby("school_name")\
                ["reading_score"].mean()
# find average reading score per school for 12th grade
grade12_read = student_data.loc[student_data["grade"] == "12th"].groupby("school_name")\
                ["reading_score"].mean()

# create dataframe
read_scores_grade = pd.DataFrame({
    "9th": grade9_read,
    "10th": grade10_read,
    "11th": grade11_read,
    "12th": grade12_read})

# format dataframe for cleaner look
read_scores_grade["9th"] = read_scores_grade["9th"].map("{:.2f}".format)
read_scores_grade["10th"] = read_scores_grade["10th"].map("{:.2f}".format)
read_scores_grade["11th"] = read_scores_grade["11th"].map("{:.2f}".format)
read_scores_grade["12th"] = read_scores_grade["12th"].map("{:.2f}".format)
read_scores_grade.index.name = None

read_scores_grade

9th   10th   11th   12th
Bailey High School     81.30  80.91  80.95  80.91
Cabrera High School    83.68  84.25  83.79  84.29
Figueroa High School   81.20  81.41  80.64  81.38
Ford High School       80.63  81.26  80.40  80.66
Griffin High School    83.37  83.71  84.29  84.01
Hernandez High School  80.87  80.66  81.40  80.86
Holden High School     83.68  83.32  83.82  84.70
Huang High School      81.29  81.51  81.42  80.31
Johnson High School    81.26  80.77  80.62  81.23
Pena High School       83.81  83.61  84.34  84.59
Rodriguez High School  80.99  80.63  80.86  80.38
Shelton High School    84.12  83.44  84.37  82.78
Thomas High School     83.73  84.25  83.59  83.83
Wilson High School     83.94  84.02  83.76  84.32
Wright High School     83.83  83.81  84.16  84.07

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [10]:
# create bins in which to place values
bins_score = [0, 585, 630, 645, 680]
# create bin labels
bin_labels_score = ["<$585", "$585-630", "$630-645", "$645-680"]
# creating new column for newly created bins
school_summary["Spending Ranges (per Student)"] = \
                pd.cut(school_summary["Per Student Budget"], bins_score, labels=bin_labels_score)
# grouped dataframe by bins
by_spending = school_summary.groupby(["Spending Ranges (per Student)"])
# create new values for new dataframe
avg_math_bin_score = by_spending["Avg Math Score"].mean()
avg_read_bin_score = by_spending["Avg Reading Score"].mean()
pass_math_bin_score = by_spending["% Passing Math"].mean()
pass_read_bin_score = by_spending["% Passing Reading"].mean()
overall_pass_bin_score = by_spending["% Overall Passing"].mean()
# creat dataframe
by_spending_df = pd.DataFrame({
    "Avg Math Score": avg_math_bin_score,
    "Avg Reading Score": avg_read_bin_score,
    "% Passing Math": pass_math_bin_score,
    "% Passing Reading": pass_read_bin_score,
    "% Overall Passing": overall_pass_bin_score})
# format cells for cleaner look
by_spending_df["Avg Math Score"] = by_spending_df["Avg Math Score"].map("{:.2f}".format)
by_spending_df["Avg Reading Score"] = by_spending_df["Avg Reading Score"].map("{:.2f}".format)
by_spending_df["% Passing Math"] = by_spending_df["% Passing Math"].map("{:.2%}".format)
by_spending_df["% Passing Reading"] = by_spending_df["% Passing Reading"].map("{:.2%}".format)
by_spending_df["% Overall Passing"] = by_spending_df["% Overall Passing"].map("{:.2%}".format)

by_spending_df

Avg Math Score Avg Reading Score % Passing Math  \
Spending Ranges (per Student)                                                   
<$585                                  83.46             83.93         93.46%   
$585-630                               81.90             83.16         87.13%   
$630-645                               78.52             81.62         73.48%   
$645-680                               77.00             81.03         66.16%   

                              % Passing Reading % Overall Passing  
Spending Ranges (per Student)                                      
<$585                                    96.61%            90.37%  
$585-630                                 92.72%            81.42%  
$630-645                                 84.39%            62.86%  
$645-680                                 81.13%            53.53%

## Scores by School Size

* Perform the same operations as above, based on school size.

In [11]:
# create bins in which to place values
bins_size = [0, 1000, 2000, 5000]
# create bin labels
bin_labels_size = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]
# creating new column for newly created bins
school_summary["School Size"] = \
                pd.cut(school_summary["Total Students"], bins_size, labels=bin_labels_size)
# grouped dataframe by bins
by_size = school_summary.groupby(["School Size"])
# create new values for new dataframe
avg_math_bin_size = by_size["Avg Math Score"].mean()
avg_read_bin_size = by_size["Avg Reading Score"].mean()
pass_math_bin_size = by_size["% Passing Math"].mean()
pass_read_bin_size = by_size["% Passing Reading"].mean()
overall_pass_bin_size = by_size["% Overall Passing"].mean()
# creat dataframe
by_size_df = pd.DataFrame({
    "Avg Math Score": avg_math_bin_size,
    "Avg Reading Score": avg_read_bin_size,
    "% Passing Math": pass_math_bin_size,
    "% Passing Reading": pass_read_bin_size,
    "% Overall Passing": overall_pass_bin_size})
# format dataframe for cleaner look
by_size_df["Avg Math Score"] = by_size_df["Avg Math Score"].map("{:.2f}".format)
by_size_df["Avg Reading Score"] = by_size_df["Avg Reading Score"].map("{:.2f}".format)
by_size_df["% Passing Math"] = by_size_df["% Passing Math"].map("{:.2%}".format)
by_size_df["% Passing Reading"] = by_size_df["% Passing Reading"].map("{:.2%}".format)
by_size_df["% Overall Passing"] = by_size_df["% Overall Passing"].map("{:.2%}".format)

by_size_df

Avg Math Score Avg Reading Score % Passing Math  \
School Size                                                          
Small (<1000)               83.82             83.93         93.55%   
Medium (1000-2000)          83.37             83.86         93.60%   
Large (2000-5000)           77.75             81.34         69.96%   

                   % Passing Reading % Overall Passing  
School Size                                             
Small (<1000)                 96.10%            89.88%  
Medium (1000-2000)            96.79%            90.62%  
Large (2000-5000)             82.77%            58.29%

## Scores by School Type

* Perform the same operations as above, based on school type

In [12]:
# grouped dataframe by School Type
by_type = school_summary.groupby(["School Type"])
# create new values for new dataframe
avg_math_bin_type = by_type["Avg Math Score"].mean()
avg_read_bin_type = by_type["Avg Reading Score"].mean()
pass_math_bin_type = by_type["% Passing Math"].mean()
pass_read_bin_type = by_type["% Passing Reading"].mean()
overall_pass_bin_type = by_type["% Overall Passing"].mean()
# creat dataframe
by_type_df = pd.DataFrame({
    "Avg Math Score": avg_math_bin_type,
    "Avg Reading Score": avg_read_bin_type,
    "% Passing Math": pass_math_bin_type,
    "% Passing Reading": pass_read_bin_type,
    "% Overall Passing": overall_pass_bin_type})
# format cells for cleaner look
by_type_df["Avg Math Score"] = by_type_df["Avg Math Score"].map("{:.2f}".format)
by_type_df["Avg Reading Score"] = by_type_df["Avg Reading Score"].map("{:.2f}".format)
by_type_df["% Passing Math"] = by_type_df["% Passing Math"].map("{:.2%}".format)
by_type_df["% Passing Reading"] = by_type_df["% Passing Reading"].map("{:.2%}".format)
by_type_df["% Overall Passing"] = by_type_df["% Overall Passing"].map("{:.2%}".format)

by_type_df

Avg Math Score Avg Reading Score % Passing Math % Passing Reading  \
School Type                                                                     
Charter              83.47             83.90         93.62%            96.59%   
District             76.96             80.97         66.55%            80.80%   

            % Overall Passing  
School Type                    
Charter                90.43%  
District               53.67%

# Trend Analysis

## Trend #1 
* District schools (which all had enrolment over 2500 students) reported much lower percentage of students passing math than those of the charter schools (which all had enrolment under 2500 students). District schools only reported ~66% of their students passing math versus over 90% of the students in the charter schools. 
* The percentage of students passing reading were a little closer between the two school types with district schools having ~80% of the students passing versus over 95% of all charter students passing reading.
* The overall number of students passing both math and reading were also vastly different between the different school types. District schools only have ~50% of their students passing both reading and math where are charter schools report ~90% of student passing both subjects.
* Another interesting bit of information is that the average scores for math (Charter 83.47 vs District 76.96) and reading (Charter 83.9 vs 80.97) are much more similar between the two school types, whereas the percentages of students passing were drastically different. This tells us that within district schools there is a large disparity in students and how well they are performing.

## Trend #2
* Another interesting trend that I noticed was that schools that spend less money per student performed much better than the school who spent more. The schools spending < 585 dollars per student (lowest bin) had the highest number of students passing each (and both) subjects (90\%) whereas the schools spending between 645- 680 dollars (highest bin) had the least number of students passing.
* I would have expected that to be opposite of what we found out.